Assets/Scripts/TextDecoder/ActionDecoder.cs

```diff
@@ -0,0 +1,416 @@
+
+        if (actionNameAndParameters.Length > 2)
+        {
+            throw new TextDecoder.Parser.ScriptParsingException($"More than one '{actionSideSeparator}' detected in line '{actionLine}'");
+        }
+
+        string action = actionNameAndParameters[0];
+        string[] parameters = (actionNameAndParameters.Length == 2) ? actionNameAndParameters[1].Split(actionParameterSeparator) : Array.Empty<string>();
+
+        // Find method with exact same name as action inside script
+        MethodInfo method = GetType().GetMethod(action, BindingFlags.Instance | BindingFlags.NonPublic);
+        if (method == null)
+        {
+            throw new TextDecoder.Parser.ScriptParsingException($"DirectorActionDecoder contains no method named '{action}'");
+        }
+
+        ParameterInfo[] methodParameters = method.GetParameters();
+        var optionalParameters = methodParameters.Count(parameter => parameter.IsOptional);
+        if (parameters.Length < (methodParameters.Length - optionalParameters) || parameters.Length > (methodParameters.Length))
+        {
+            throw new TextDecoder.Parser.ScriptParsingException($"'{action}' requires {(optionalParameters == 0 ? "exactly" : "between")} {(optionalParameters == 0 ? methodParameters.Length.ToString() : $"{methodParameters.Length-optionalParameters} and {methodParameters.Length}")} parameters (has {parameters.Length} instead)");
+        }
+
+        List<object> parsedMethodParameters = new List<object>();
+        // For each supplied parameter of that action...
+        for (int index = 0; index < parameters.Length; index++)
+        {
+            if (parameters.Length <= index && methodParameters[index].IsOptional)
+            {
+                parsedMethodParameters.Add(methodParameters[index].DefaultValue);
+            }
+
+            // Determine it's type
+            ParameterInfo methodParameter = methodParameters[index];
+
+            // Edge-case for enums
+            if (methodParameter.ParameterType.BaseType == typeof(Enum))
+            {
+                try
+                {
+                    parsedMethodParameters.Add(Enum.Parse(methodParameter.ParameterType, parameters[index]));
+                    continue;
+                }
+                catch (ArgumentException e)
+                {
+                    Regex pattern = new Regex(@"Requested value '(.*)' was not found\.");
+                    Match match = pattern.Match(e.Message);
+                    if (match.Groups.Count > 0)
+                    {
+                        throw new TextDecoder.Parser.ScriptParsingException($"'{parameters[index]}' is incorrect as parameter #{index + 1} ({methodParameter.Name}) for action '{action}': Cannot convert '{match.Groups[1].Captures[0]}' into an {methodParameter.ParameterType} (valid values include: '{string.Join(", ", Enum.GetValues(methodParameter.ParameterType).Cast<object>().Select(a=>a.ToString()))}')");
```

Comment:
```suggestion
                    Match patternMatch = new Regex(@"Requested value '(.*)' was not found\.").Match(e.Message);
                    if (patternMatch.Groups.Count > 0)
                    {
                        throw new TextDecoder.Parser.ScriptParsingException($"'{parameters[index]}' is incorrect as parameter #{index + 1} ({methodParameter.Name}) for action '{action}': Cannot convert '{patternMatch.Groups[1].Captures[0]}' into an {methodParameter.ParameterType} (valid values include: '{string.Join(", ", Enum.GetValues(methodParameter.ParameterType).Cast<object>().Select(a=>a.ToString()))}')");
```
Can this be chained into one line? Or does the Regex need to be instantiated in its own variable?